### PR TITLE
Challenges

### PR DESCRIPTION
**Pull request by omarradwan**
active da walla msh active :'D

### CODE REVIEW

**Comment by omarradwan** [comment]
shof ba2a ya sagheer 3shan a5od user story kman B|

File: app/Http/Controllers/VolunteerController.php [hunk 1]
```diff
@@ -2,12 +2,13 @@
...
 namespace App\Http\Controllers;
 
+use App\Challenge;
 use Illuminate\Http\Request;
 use App\Http\Requests;
 use App\User;
 
 use App\Event;
-use App\Http\Requests;
+use Illuminate\Support\Facades\Auth;
```
Comment (AhmadElsagheer) [inline app/Http/Controllers/VolunteerController.php:14]: Facades can be used as `use [Facade_name];`

File: app/Http/Controllers/VolunteerController.php [hunk 2]
```diff
@@ -60,5 +61,65 @@ public static function show($id)
...
+        if($challenge->user->id == Auth::user()->id)
+        {
+            $challenge->update($request->all());
+            return redirect()->action('VolunteerController@show' , [Auth::user()->id]);
+        }
+        return redirect('/');
+
+    }
+
+
```
Comment (AhmadElsagheer) [inline app/Http/Controllers/VolunteerController.php]: when creating a function, always specify the authenticated users in the middlewares. In this case, these functions are allowed for volunteers only

File: app/Http/Controllers/VolunteerController.php [hunk 3]
```diff
@@ -60,5 +61,65 @@ public static function show($id)
...
+    /**
+     *  Store the challenge in the database
+     */
+    public function storeChallenge(Request $request)
+    {
+        $this->validate($request ,
+            ['events' => 'required|numeric|min:1' ,
+             'deadline' => 'required|date_format:Y-d-m'
+            ]);
+        $user_id = Auth::user()->id;
```
Comment (AhmadElsagheer) [inline app/Http/Controllers/VolunteerController.php]: the challenges are designed for a whole year (similar to that implement in goodreads). That is, the volunteer is promoted to set a number of events to attend in the current year

**Comment by omarradwan** [comment]
In the previous commit, I designed it to be for a whole year. But the user story stated that a user can set/edit a challenge to himself concerning events to join in a certain time. It did not specify the period. So I left that option to the user to set the deadline himself.
lw msh 3agbak nerga3 lel commit elly ablo bra7tak :'D

File: app/Http/routes.php [hunk 4]
```diff
@@ -151,6 +151,17 @@
...
+     *  set challenges
+     */
+    Route::get('challenge' , 'VolunteerController@createChallenge');
+    Route::post('challenge' , 'VolunteerController@storeChallenge');
+
+    /**
+     *  edit challenges
+     */
+    Route::get('challenge/{challenge_id}/edit' , 'VolunteerController@editChallenge');
+    Route::patch('challenge/{challenge_id}/edit' , 'VolunteerController@updateChallenge');
```
Comment (AhmadElsagheer) [inline app/Http/routes.php:166]: must be added before the resource

**Comment by omarradwan** [comment]
why?

**Comment by AhmadElsagheer** [comment]
Good job Omar. I have made some small edits only. The user can edit the challenge of the current year only. Regarding the routes, it is written in the documentation that any extra routes must be before the resource. The problem didn't appear when the route was `challenge\`. I edited it to be `volunteer\challenge` to be more specific and here this ordering is important. Bs 3ash wallahy :+1: :+1:  Now I want you to work on a view that will consist of two parts. The first part is showing the current year challenge with a button to edit this challenge that will redirect to the edit view. The second part is a listing of past challenges. You can make them in one page one after another. If you managed to make a graph for the past challenges with the current achieved events in the current year, that will be great :D. Yes, you can notice, we forget to make a column for the current attended events :"D You can edit the migration to make have goal_events and attended_events. Also, find where the attended event must be updated (hint: it will be in a method in EventController)

**Pull request merged**